packages/drafts-realm/leaflet-gtfs.gts

```diff
@@ -0,0 +1,256 @@
+
+  @field lat = contains(NumberField);
+  @field lon = contains(NumberField);
+
+  @field tileserverUrl = contains(StringField);
+  @field gtfsUrl = contains(StringField);
+
+  map: any;
+
+  @tracked routes: Route[] = [];
+  @tracked activeRouteId: string | null = null;
+
+  @tracked shapes: any[] = [];
+
+  @tracked trips: any[] = [];
+
+  @tracked routesFeatureGroup: any = null;
+  @tracked routePolylines: any[] = [];
+
+  @action
+  setMap(map: any) {
+    this.map = map;
+  }
+
+  @action
+  setRoutes(routes: any) {
+    this.routes = routes.map((routeCsv: any) => ({
+      id: routeCsv.route_id,
+      name: `${routeCsv.route_short_name} ${routeCsv.route_long_name}`,
+    }));
+  }
+
+  @action
+  setShapes(shapes: any) {
+    this.shapes = shapes;
+  }
+
+  @action
+  setTrips(trips: any) {
+    this.trips = trips;
+  }
+
+  @action
+  handleRouteSelect(e: Event) {
+    if (e.target) {
+      let select = e.target as HTMLSelectElement;
+      this.setActiveRouteId(select.value);
+    }
+  }
+
+  @action
+  setActiveRouteId(activeRouteId: string) {
+    this.activeRouteId = activeRouteId;
+
+    this.routesFeatureGroup?.remove();
+
+    let activeRouteVariantPoints = this.activeRouteVariantPoints;
+
+    if (activeRouteVariantPoints) {
+      this.routesFeatureGroup = L.featureGroup(
+        activeRouteVariantPoints.map((points) => {
+          return L.polyline(points);
+        }),
+      );
+
+      this.routesFeatureGroup.addTo(this.map);
+      this.map.fitBounds(this.routesFeatureGroup.getBounds());
+    } else {
+      this.routesFeatureGroup = null;
+    }
+  }
+
+  get activeRouteShapeIds() {
+    if (this.trips) {
+      return this.trips.reduce((shapeIds, trip) => {
+        if (trip.route_id === this.activeRouteId) {
+          shapeIds.add(trip.shape_id);
+        }
+        return shapeIds;
+      }, new Set());
+    }
+  }
+
+  get activeRouteVariantPoints() {
+    if (this.shapes && this.activeRouteShapeIds) {
+      let activeRouteShapeIds = this.activeRouteShapeIds;
+
+      let shapeIdToPoints = new Map<string, [number, number][]>();
+
+      this.shapes.forEach((shape) => {
+        if (activeRouteShapeIds.has(shape.shape_id)) {
+          if (!shapeIdToPoints.has(shape.shape_id)) {
+            shapeIdToPoints.set(shape.shape_id, []);
+          }
+
+          shapeIdToPoints
+            ?.get(shape.shape_id)
+            ?.push([shape.shape_pt_lat, shape.shape_pt_lon]);
+        }
+      });
+
+      return Array.from(shapeIdToPoints.values());
+    }
+
+    return [];
+  }
+
+  static isolated = class Isolated extends Component<typeof this> {
+    <template>
+      <figure
+        {{LeafletModifier
+          lat=@model.lat
+          lon=@model.lon
+          tileserverUrl=@model.tileserverUrl
+          setMap=@model.setMap
+        }}
+        {{GtfsModifier
+          gtfsUrl=@model.gtfsUrl
+          setRoutes=@model.setRoutes
+          setShapes=@model.setShapes
+          setTrips=@model.setTrips
```

Comment:
These callbacks all produce [type warnings](https://github.com/cardstack/boxel/actions/runs/9229440201/job/25395653675?pr=1241#step:14:31). Why would `@model.setTrips` ever be undefined?

Fetching a GTFS archive, unzipping, and processing it is maybe a questionable use of a modifier, what’s a better approach?

Reply:
I wonder if it's because the type for the `@model` is `Partial<LeafletGtfs>`. Maybe it's not sure that the `setTrips` method will be on the model, even when the modifier signature is adjusted to not be optional. I don't know why the `@model` type is defined as a partial in card-api.